Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+      MergerDigit& d2 = previousBuffer.digits[j];
+      if (checkDigits(d1, d2)) {
+        break;
+      }
+    }
+  }
+}
+
+void Merger::setOrbit(int feeId, uint32_t orbit, bool stop)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].setOrbit(orbit, stop);
+}
+
+void Merger::setDigitHandler(std::function<void(const Digit&)> h)
+{
+  for (int feeId = 0; feeId <= MCH_MERGER_FEEID_MAX; feeId++) {
+    mergers[feeId].setDigitHandler(h);
+  }
+}
+
+void Merger::addDigit(int feeId, int solarId, int dsAddr, int chAddr,
+                      int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].getCurrentBuffer().digits.emplace_back(MergerDigit{o2::mch::Digit(deId, padId, adc, time),
+                                                                    stopTime, false, solarId, dsAddr, chAddr});
+}
+
+void Merger::mergeDigits(int feeId)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].mergeDigits();
+}
+
+//=======================
+// Data decoder
+void DataDecoder::decodeBuffer(gsl::span<const std::byte> page)
+{
+  size_t ndigits{0};
+
+  uint8_t isStopRDH = 0;
+  uint32_t orbit;
+  uint32_t feeId;
+  uint32_t linkId;
+
+  const auto storeDigit = [&](const Digit& d) {
+    mOutputDigits.emplace_back(d);
+    if (mPrint) {
+      std::cout << "[storeDigit]: digit stored, mOutputDigits size: " << mOutputDigits.size() << std::endl;
+    }
+  };
+
+  auto channelHandler = [&](DsElecId dsElecId, uint8_t channel, o2::mch::raw::SampaCluster sc) {
+    if (mChannelHandler) {
+      mChannelHandler(dsElecId, channel, sc);
+    }
+
+    if (mDs2manu) {
+      channel = ds2manu(int(channel));
+    }
+
+    double digitadc(0);
+    for (auto& s : sc.samples) {
+      digitadc += s;
+    }
+
+    int deId{-1};
+    int dsIddet{-1};
+    if (auto opt = mElec2Det(dsElecId); opt.has_value()) {
+      DsDetId dsDetId = opt.value();
+      dsIddet = dsDetId.dsId();
+      deId = dsDetId.deId();
+    }
+
+    int padId = -1;
+    const Segmentation& segment = segmentation(deId);
+    if ((&segment) == nullptr) {
+      return;
+    }
+
+    padId = segment.findPadByFEE(dsIddet, int(channel));
+    if (mPrint) {
+      auto s = asString(dsElecId);
+      auto ch = fmt::format("{}-CH{:02d}", s, channel);
+      std::cout << ch << "  "
+                << fmt::format("PAD ({:04d} {:04d} {:04d})\tADC {:5.0f}  TIME ({} {} {:02d})  SIZE {}  END {}",
+                               deId, dsIddet, padId, digitadc, orbit, sc.bunchCrossing, sc.sampaTime, sc.nofSamples(), (sc.sampaTime + sc.nofSamples() - 1))
+                << (((sc.sampaTime + sc.nofSamples() - 1) >= 98) ? " *" : "") << std::endl;
+    }
+
+    // skip channels not associated to any pad
+    if (padId < 0) {
+      return;
+    }
+
+    Digit::Time time;
```

Comment:
here again are opportunities to a couple free functions to make the code clearer ? 
e.g. 
```
Digit::Time startTime(const SampaCluster& sc, uint32_t orbit) { return Digit::Time{sc.sampaTime,sc.bunchCrossing,orbit); }
Digit::Time stopTime(const SampaCluster& sc, uint32_t orbit) { return Digit::Time{sc.sampaTime+sc.nofSamples()-1,sc.bunchCrossing,orbit); }
```

Reply:
Again, I would propose to wait for a possible improved definition of the digit time based on the bunch crossing value at the beginning of each Time Frame.

Reply:
well, seeing that actually the only thing needed for the client code (looking here at DigitsMerger.cxx L121) is the timeDiff, I would write a timeDiff(const Digit& d1, const Digit& d2) and let the implementation of that timeDiff change if need be later on (including the removal of the printouts at some point) ?  what do you think ?

Reply:
Good, point, I have introduced the timeDiff function as you propose.